File: frontend/src/api/external-asset-data/naptan-data-source-handler.ts
import type { Feature, GeoJsonProperties, Geometry } from "geojson";
import Papa from "papaparse";
import { DataSourceHandler } from "./data-source-handler";
import { AssetSpecification } from "@/hooks/queries/dataset-utils";

interface NaptanStop {
  longitude: number;
  latitude: number;
  ATCOCode: string;
  [key: string]: any;
}

export class NaptanDataSourceHandler extends DataSourceHandler {
  /**
   * A function to build a list of NAPTAN URLs.
   *
   * @param assetSpecification the specification of the asset
   * @returns
   */
  public buildUrlsForDataSource(
    _assetSpecification: AssetSpecification,
  ): string[] {
    return [
      `/transparent-proxy/naptan/v1/access-nodes?dataFormat=csv&atcoAreaCodes=${this.locator}`,
    ];
  }

  /**
   * Fetches CSV data from a given URL and parses it.
   *
   * @param url A URL
   * @returns Parsed CSV data as a list of {@link NaptanStop}
   */
  async fetchCsvFromUrl(url: string) {
    return await fetch(url)
      .then((res) => res.text())
      .then((csvText) => {
        return Papa.parse<NaptanStop>(csvText, { header: true });
      });
  }

  /**
   * A function which fetches all data from the OS NGD API for a given asset specification.
   *
   * @param assetSpecification the specification of the asset
   * @param url the URL to be queried for data
   * @returns a list of mapped GeoJSON-compatible features
   */
  public async fetchDataForAssetSpecification(
    assetSpecification: AssetSpecification,
    url: string,
  ): Promise<Feature[]> {
    const features: Feature[] = [];
    const parsedData = await this.fetchCsvFromUrl(url);
    const stops = parsedData.data;
    for (const stop of stops) {
      features.push(this.createGeoJsonPointFeature(stop));
    }
    const seenLocations = new Set<string>();
    const featureForAssetType = features.filter((f) =>
      this.isMatchForAssetSpecificationFilters(assetSpecification, f),
    );
    return featureForAssetType.reduce<Feature[]>(
      (acc, f) =>
        this.mergeAssetsWithSameNameAtSameLocation(acc, f, seenLocations),
      [],
    );
  }

  /**
   * Evaluates whether a feature has already been marked for inclusion and if not pushes it to a list.
   *
   * @param featuresToInclude A list of features which have been marked for inclusion.
   * @param featureToEvaluate A featture to evaluate for inclusion.
   * @param seenLocations A list of locations which have already been seen.
   * @returns
   */
  private mergeAssetsWithSameNameAtSameLocation(
    featuresToInclude: Feature<Geometry, GeoJsonProperties>[],
    featureToEvaluate: Feature<Geometry, GeoJsonProperties>,
    seenLocations: Set<string>,
  ) {
    if (!featureToEvaluate.properties?.name) {
      featuresToInclude.push({ ...featureToEvaluate });
    } else {
      const location = `${featureToEvaluate.properties.name}_${featureToEvaluate.properties.LocalityName}`;
      if (seenLocations.has(location)) {
        return featuresToInclude;
      }
      seenLocations.add(location);
      featuresToInclude.push({ ...featureToEvaluate });
    }
    return featuresToInclude;
  }

  /**
   * Creates a GeoJSON feature from a given data point.
   * @param stop an instance of {@link NaptanStop}
   * @returns
   */
  private createGeoJsonPointFeature(stop: NaptanStop): Feature {
    const {
      Longitude: longitude,
      Latitude: latitude,
      ATCOCode: atcoCode,
      CommonName: name,
      ...properties
    } = stop;

    return {
      id: atcoCode,
      type: "Feature",
      geometry: {
        type: "Point",
        coordinates: [
          this.convertToNumberIfString(longitude),
          this.convertToNumberIfString(latitude),
        ],
      },
      properties: {
        ...properties,
        name,
      },
    };
  }

  private convertToNumberIfString(val: number | string): number {
    return typeof val === "string" ? Number(val) : val;
  }
}

File: backend/vista-python-api/src/api/repository/external/asset_data_source_handlers/naptan_data_source_handler.py
# SPDX-License-Identifier: Apache-2.0
# © Crown Copyright 2026. This work has been developed by the National Digital Twin Programme and is legally attributed to the Department for Business and Trade (UK) as the governing entity.

"""Handler for the NAPTAN data source."""

from functools import reduce

from api.repository.external.external_asset_mapper import ExternalAssetMapper

from .base import DataSourceHandler


class NaptanDataSourceHandler(DataSourceHandler):
    """Handler for the NAPTAN data source."""

    def build_urls_for_data_source(self, _asset_specification):
        """Build the URLs for fetching the data per the specification given."""
        return [
            f"https://naptan.api.dft.gov.uk/v1/access-nodes?dataFormat=csv&atcoAreaCodes={self.locator}"
        ]

    async def fetch_data_for_asset_specification(self, asset_specification, url):
        """Fetch the OS NGD data per the specification given."""
        stops = await self.fetch_csv_from_url(url)
        unique_stops = reduce(self._merge_stops_with_same_name_at_same_location(set()), stops, [])
        return [
            ExternalAssetMapper.map_from_naptan(stop, asset_specification)
            for stop in unique_stops
            if self._is_stop_match_for_asset_specification_filters(
                stop, asset_specification["filters"]
            )
        ]

    def _merge_stops_with_same_name_at_same_location(self, seen):
        def reduce(stops, next_stop):
            if "CommonName" not in next_stop:
                stops.append(next_stop)
            else:
                location = f"{next_stop['CommonName']}_{next_stop['LocalityName']}"
                if location in seen:
                    return stops
                seen.add(location)
                stops.append(next_stop)
            return stops

        return reduce

    def _is_stop_match_for_asset_specification_filters(self, stop, data_filters):
        if data_filters:
            conditions = []
            for data_filter in data_filters:
                if isinstance(data_filter["filterValue"], list):
                    conditions.append(
                        any(
                            stop[f"{data_filter['filterName']}"] == filter_value
                            for filter_value in data_filter["filterValue"]
                        )
                    )
                else:
                    conditions.append(
                        stop[f"{data_filter['filterName']}"] == data_filter["filterValue"]
                    )
            return all(conditions)
        return True
